analytics/location_heatmaps/geo_utils.py

```diff
@@ -276,3 +435,167 @@ def quantize_vector(vector, left_bound, right_bound):
   scale = (vector - left_bound) // distance
   vector -= distance * scale
   return vector
+
+
+def makeGaussian(image, total_size, fwhm=3, center=None,
+                 convert=False, save=False, load=False):
+  """ Make a square gaussian kernel.
+  size is the length of a side of the square
+  fwhm is full-width-half-maximum, which
+  can be thought of as an effective radius.
+  """
+  import torch
+
+  if load:
+    return torch.load(f'split_dataset_{fwhm}_{center[0]}_{center[1]}.pt')
+  size = image.shape[0]
+  x = np.arange(0, size, 1, float)
+  y = x[:, np.newaxis]
+
+  if center is None:
+    x0 = y0 = size // 2
+  else:
+    x0 = center[0]
+    y0 = center[1]
+  hotspot = np.exp(-4 * np.log(2) * ((x - x0) ** 2 + (y - y0) ** 2) / fwhm ** 2)
+  pos_image = np.floor(hotspot * image)
+  pos_image = pos_image.astype(int)
+  neg_image = image - pos_image
+
+  if convert:
+    pos_dataset = convert_to_dataset(pos_image, total_size, value=1)
+    neg_dataset = convert_to_dataset(neg_image, total_size, value=0)
+    total_dataset = np.concatenate([pos_dataset, neg_dataset])
+    res = dict(mask=hotspot, pos_image=pos_image, neg_image=neg_image,
+               pos_dataset=pos_dataset, neg_dataset=neg_dataset,
+               total_dataset=total_dataset)
+    if save:
+      torch.save(res, f'split_dataset_{fwhm}_{center[0]}_{center[1]}.pt')
+      print(f'Saved to split_dataset_{fwhm}_{center[0]}_{center[1]}.pt')
+    return res
+  else:
+    return dict(mask=hotspot, pos_image=pos_image, neg_image=neg_image)
+
+
+def convert_to_dataset(image, total_size, value=None):
+  if value is not None:
+    dataset = np.zeros(image.sum(),
+                       dtype=[('x', np.int16), ('y', np.int16),
+                              ('pos', np.int8)])
+  else:
+    dataset = np.zeros(image.sum(),
+                       dtype=[('x', np.int16), ('y', np.int16)])
+  z = 0
+  for i in tqdm(range(total_size), total=total_size):
+    for j in range(total_size):
+      for _ in range(int(image[i, j])):
+        if value is not None:
+          dataset[z] = (i, j, value)
+        else:
+          dataset[z] = (i, j)
+        z += 1
+
+  return dataset
+
+
+def compute_conf_intervals(sum_vector: np.ndarray, level=95):
+  conf_intervals = dict()
+  conf_interval_weighted = dict()
+  if level == 95:
+    z = 1.96
+  elif level == 99:
+    z = 2.576
+  elif level == 90:
+    z = 1.645
+  elif level == 98:
+    z = 2.326
+  else:
+    raise ValueError(f'Incorrect confidence level {level}.')
```

Comment:
It'd be nicer to just compute the z score analytically, rather than having sparse lookup table. I think the following should do the trick:

from scipy.stats import norm

z = norm.ppf(1-(1-level/100)/2)

Reply:
updated